lib/src/generator/file_structure.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:dartdoc/src/model/model_element.dart';
+
+enum FileStructureMode {
+  htmlOriginal,
+  mdOriginal,
+}
+
+/// This class defines an interface to allow [ModelElement]s and [Generator]s
+/// to get information about the desired on-disk representation of a single
+/// [ModelElement].  None of these getters should be considered valid unless
+/// [modelElement.isCanonical] is true.  Implementations of this class can
+/// define which elements have their own files and which do not, how
+/// to lay them out on disk, and how to link different pages in the structure
+/// together.
+abstract class FileStructure {
+  factory FileStructure(FileStructureMode mode, ModelElement modelElement) {
+    switch (mode) {
+      case FileStructureMode.htmlOriginal:
+        return _FileStructureHtml(modelElement);
+      case FileStructureMode.mdOriginal:
+        return _FileStructureMd(modelElement);
+    }
+  }
+
+  /// Link to the [ModelElement] the information for this [FileStructure]
+  /// applies to.
+  ModelElement get modelElement;
```

Comment:
Do you know why this would be needed? I Or need to be public? It would be good to instantiate the fields at construction time and not need to hold onto the `modelElement`.

Reply:
Some of the information here (particularly on whether or not a file needs to be created) depends on canonicalization information and not all of that is available at construction time for the ModelElement object.  Since I was planning to construct the `FileStructure` objects at or very near `ModelElement` construction time, I was thinking that holding on to the modelElement would avoid construction loops.   But maybe with deferred construction we can avoid that?  I will add a todo.

Reply:
Another possibility is we can just calculate all this stuff as though we are canonical (even if we aren't) and avoid construction loops that way.   That has implications for your other question, above.

Reply:
Hard to avoid YAGNI designing an API before it is used. Hard to review as well.